### PR TITLE
Add back-end's internal state migration support

### PR DESCRIPTION
**Pull request by germag**
### Summary of the PR
  
A new set of messages have been introduced in the vhost-user protocol to support the migration of devices with back-end's internal state[0].
    
These are the additions to the protocol:
    - New vhost-user protocol feature `VHOST_USER_PROTOCOL_F_DEVICE_STATE`
    - `SET_DEVICE_STATE_FD` function: Front-end and back-end negotiate a file
      descriptor over which to transfer the state. It also includes
      establishing the direction of transfer and migration phase.
    - `CHECK_DEVICE_STATE`: After the state has been transferred through the
      file descriptor, the front-end invokes this function to verify
      success.  There is no in-band way (through the file descriptor) to
      indicate failure, so we need to check explicitly.
    
[0] https://github.com/qemu/qemu/commit/019233096c03b826e0e677115b6e3c550a54a48d

This code is part of the effort to support live migration in virtiofsd and is currently being used in a version of virtiofsd that supports migration: https://gitlab.com/virtiofsd-live-migration

### Note to reviewers

~Please do not merge this PR yet, this PR depends on https://github.com/rust-vmm/vhost/pull/202, because by adding `FrontendReq::SET_DEVICE_STATE_FD` a "hole" was created in value `41`, this makes the APi unsound, since `VhostUserMsgHeader::get_code()` call `transmute_copy()` which assumes that the values are continuous and have no "holes" between them.~

~I'll rebase and adjust the code after PR 202 gets merged.~

### CODE REVIEW

File: crates/vhost-user-backend/src/backend.rs
Comment (XanClic) [inline crates/vhost-user-backend/src/backend.rs:121]: Maybe we should add documentation on the return type; without reading the vhost-user doc, it isn’t obvious why a `File` can be returned. Something like “The back-end can choose to use a different channel than `file`. If so, it must return a `File` that the front-end can use.”?

File: crates/vhost-user-backend/src/backend.rs
Comment (XanClic) [inline crates/vhost-user-backend/src/backend.rs:489]: Again my fault (like the missing comment), but won’t `read()` suffice here? `VhostUserBackendMut::check_device_state()` only takes an immutable `&self`.

File: crates/vhost-user-backend/src/backend.rs
Comment (XanClic) [inline crates/vhost-user-backend/src/backend.rs:139]: Personally, I liked using `std::result::Result<(), ()>` to make it explicit in typing that the error message would be lost, because the function/message name sounds like the front-end might want to inquire about the exact nature of the error (and documentation is easy to miss).
But then again, most messages allow you to return a non-trivial error, and I don’t think any of them will return anything more than “error/success” to the front-end, so really, there is no point in treating this any differently from those other messages.

File: crates/vhost/src/vhost_user/backend_req_handler.rs
Comment (XanClic) [inline crates/vhost/src/vhost_user/backend_req_handler.rs:587]: As far as I can see, `send_ack_message()` only sends a reply if `F_REPLY_ACK` has been negotiated. However, `CHECK_DEVICE_STATE` is defined to always send a reply, independently of `F_REPLY_ACK`. So I don’t think we should use `send_ack_message()` here (or if we do, I think it would need an additional parameter to force it to send the reply message).

**Comment by XanClic** [comment]
The compiled version of the documentation linked in patch 1 is here: https://qemu-project.gitlab.io/qemu/interop/vhost-user.html

File: crates/vhost-user-backend/src/backend.rs
Comment (germag) [reply crates/vhost-user-backend/src/backend.rs:121]: Yes, I plan to add more docs (for that reason I leave the [Draft])

File: crates/vhost-user-backend/src/backend.rs
Comment (germag) [reply crates/vhost-user-backend/src/backend.rs:489]: Yes, I missed that one

File: crates/vhost/src/vhost_user/backend_req_handler.rs
Comment (germag) [reply crates/vhost/src/vhost_user/backend_req_handler.rs:587]: True

File: crates/vhost-user-backend/src/backend.rs
Comment (germag) [reply crates/vhost-user-backend/src/backend.rs:139]: Nop, it's a clippy warn:
https://rust-lang.github.io/rust-clippy/master/index.html#/result_unit_err

File: crates/vhost-user-backend/src/backend.rs
Comment (XanClic) [reply crates/vhost-user-backend/src/backend.rs]: (Looks good to me in the current version. (github won’t let me resolve the conversation, because only the PR author and anyone with write access to the repo can do so, hence this comment.))

File: crates/vhost-user-backend/src/backend.rs
Comment (XanClic) [reply crates/vhost-user-backend/src/backend.rs]: Interesting, for some reason I couldn’t reproduce that warning (and didn’t find anything about it). Ah well.

File: crates/vhost/src/vhost_user/backend_req_handler.rs
Comment (XanClic) [reply crates/vhost/src/vhost_user/backend_req_handler.rs]: (Looks good in the current version.)

**Comment by XanClic** [comment]
Looks good to me, thanks!

**Review by stefano-garzarella — APPROVED** [review]
nit: I see we are using `back end`, `back-end`, and `backend` in this PR, but also in the code that is already merged, so I think we should decide which term to use (maybe `back-end` to follow the spec) and use it all over the place (not for this PR).

File: vhost/src/vhost_user/message.rs
Comment (Ablu) [inline vhost/src/vhost_user/message.rs:939]: I would add a packed here to simplify reviews of unsafe `ByteValued` impls.

File: vhost/src/vhost_user/message.rs
Comment (Ablu) [inline vhost/src/vhost_user/message.rs:948]: POD is only one requirement. Since we seem to got this wrong in a few places we should also mention the packed requirement here IMHO. What do you think?

**Review by Ablu — COMMENTED** [review]
LGTM. Only some nit-picking on the ByteValued. I think since it has been source of confusion before, we should probably be a bit more explicit about it's requirements.

**Comment by germag** [comment]
v2:
- Add comments in message definition

File: vhost/src/vhost_user/message.rs
Comment (germag) [reply vhost/src/vhost_user/message.rs]: Done

**Comment by germag** [comment]
Rebase

**Pull request merged**
